runtime/src/main/cpp/FinalizerHooksTest.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "FinalizerHooks.hpp"
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+#include "FinalizerHooksTestSupport.hpp"
+#include "Memory.h"
+
+using namespace kotlin;
+
+using ::testing::_;
+
+namespace {
+
+class FinalizerHooksTest : public testing::Test {
+public:
+    testing::MockFunction<void(ObjHeader*)>& finalizerHook() { return finalizerHooks_.finalizerHook(); }
+
+private:
+    FinalizerHooksTestSupport finalizerHooks_;
+};
+
+} // namespace
+
+TEST_F(FinalizerHooksTest, TypeWithFinalizerHookWithoutExtra) {
+    TypeInfo type;
+    type.typeInfo_ = &type;
+    type.flags_ |= TF_HAS_FINALIZER;
+    ObjHeader obj = {&type};
+    ASSERT_FALSE(obj.has_meta_object());
+
+    EXPECT_TRUE(HasFinalizers(&obj));
+    EXPECT_CALL(finalizerHook(), Call(&obj));
+    RunFinalizers(&obj);
+    EXPECT_FALSE(obj.has_meta_object());
+
+    testing::Mock::VerifyAndClearExpectations(&finalizerHook());
```

Comment:
Btw, this looks kinda fragile. What happens if one forgets to add this line?
Can we make this implicitly?

Reply:
It's not really needed: the verification should happen automatically during mock's destructor. I'm placing them by the force of habbit.

Reply:
Well, then this might be confusing, and also probably will be copied by other developers to new tests.
So if this happens automatically at the end of the test, then it makes some sense to remove these lines.

Reply:
Good point, cleaned them up.